gammapy/maps/core.py

```diff
@@ -894,11 +894,120 @@ def interp_to_geom(self, geom, preserve_counts=False, fill_value=0, **kwargs):
 
         return Map.from_geom(geom, data=data, unit=self.unit)
 
+    def reproject(self, geom, preserve_counts=False, fill_value=0, method="oversampling", oversampling_factor=10):
+        """Reproject map to input geometry and interpolate along the non-spatial dimensions.
+
+        Parameters
+        ----------
+        geom : `~gammapy.maps.Geom`
+            Target Map geometry
+        preserve_counts : bool
+            Preserve the integral over each bin.  This should be true
+            if the map is an integral quantity (e.g. counts) and false if
+            the map is a differential quantity (e.g. intensity)
+        fill_value : None or float value
+            The value to use for points outside of the interpolation domain.
+            If None, values outside the domain are extrapolated.
+            Used only if the non-spatial axes require to be interpolated.
+        method : {'oversampling', 'polygon'}
+            Method to reproject data to a new projection
+            -oversampling: the base map is oversampled and 
+             its bins are re-distributed into a map with the target geom.
+            -polygon: flux-conserving spherical polygon intersection
+            from reproject.reproject_exact
+        oversampling_factor : int
+           Minimal factor between the bin size of the output map and the oversampled base map.
+           Used only for the oversampling method.
+
+        Returns
+        -------
+        output_map : `Map`
+            Reprojected Map
+        """
+        from .region import RegionGeom
+
+        if isinstance(geom, RegionGeom):
+            raise TypeError(f"Reproject is not supported for {type(geom)}")
+        if isinstance(self.geom, RegionGeom):
+            raise TypeError(f"Reproject is not supported for {type(self.geom)}")
+        
+        if geom.is_image:
+            axes = [ax.copy() for ax in self.geom.axes]
+            geom = geom.copy(axes=axes)
+            axes_eq = True
+        elif geom.ndim != self.geom.ndim:
+            raise TypeError("Invalid axes in the target geom.")
+        else:
+            axes_eq = np.all(
+                [ax0 == ax1 for ax0, ax1 in zip(geom.axes, self.geom.axes)]
+            )
+
+        if not axes_eq:
```

Comment:
For templates the axes will rarely be the same so it will fail often, I would prefer to return the reprojection of the spatial part keeping original axes by default and add an option interp_axes  to allow the interpolation along the axes for convenience.